File: Java/src/main/java/org/codecop/lcdnumbers/Digit.java
package org.codecop.lcdnumbers;

import java.util.List;
import java.util.Objects;

/**
 * Value Object of an LCD digit.
 */
public final class Digit {

    private final int digit;
    private final List<Line> lines;

    public Digit(int digit, Patterns patterns) {
        this(digit, patterns.of(digit));
    }

    private Digit(int digit, List<Line> lines) {
        Objects.requireNonNull(lines);

        this.digit = digit;
        this.lines = lines;
    }

    public List<Line> lines() {
        return lines;
    }

    public Digit scale(List<Line> scaledLines) {
        Objects.requireNonNull(scaledLines);

        int linesCount = lines.size();
        int scaledCount = scaledLines.size();
        if (scaledCount < linesCount) {
            throw new IllegalArgumentException("Scaled lines must be more than original ones: " + scaledCount + ">=" + linesCount);
        }

        return new Digit(digit, scaledLines);
    }

    @Override
    public boolean equals(Object other) {
        if (!(other instanceof Digit)) {
            return false;
        }
        Digit that = (Digit) other;
        return this.digit == that.digit;
    }

    @Override
    public int hashCode() {
        return digit;
    }

    @Override
    public String toString() {
        return Integer.toString(digit);
    }
}

File: Kotlin-Spring/src/main/java/org/codecop/lcdnumbers/Digit.kt
package org.codecop.lcdnumbers

import java.util.Objects

/**
 * Value Object of an LCD digit.
 */
class Digit private constructor(
    private val digit: Int,
    private val lines: List<Line>
) {
    constructor(digit: Int, patterns: Patterns) : this(digit, patterns.of(digit))
    
    fun lines(): List<Line> {
        return lines
    }
    
    fun scale(scaledLines: List<Line>): Digit {
        val linesCount = lines.size
        val scaledCount = scaledLines.size
        require(scaledCount >= linesCount) { "Scaled lines must be more than original ones: $scaledCount>=$linesCount" }
        return Digit(digit, scaledLines)
    }
    
    override fun toString(): String {
        return digit.toString()
    }
}